### ISSUE TITLE
Calling NSS_InitContext() without define-able parameters breaks the ability to use the NSS keystore

### ISSUE THREAD

**Issue by jhlange** [rpm-software-management/rpm#654]
**TLDR;**  
rpmio should call NSS_InitContext() with parameters that can be controlled from a configuration file.

The init call should allow the flags like 'NSS_INIT_NOCERTDB' to be overridden
Should allow the first argument to be specified in a configuration file.

**This is how curl calls the method**
https://github.com/curl/curl/blob/8ad9e5915a95258267f2af463559ded134e8e643/lib/vtls/nss.c#L1321

**This is how RPM call the method (prior to curl being able to set its preferred values**
https://github.com/rpm-software-management/rpm/blob/92eadae94c48928bca90693ad63c46ceda37d81f/rpmio/digest_nss.c#L64

**The issue**
RHEL uses the NSS library for SSL when building curl instead of OpenSSL, which is the underlying library responsible for downloading content in yum. 

An unfortunate side effect of this is that the parsing of many types of PEM formats and key types can not be read from PEM files. More information on that is here: https://bugzilla.redhat.com/show_bug.cgi?id=1440873

A solution to this problem is using the NSS databases to store private keys, including client certificates (in this case an EC cert). The solution works well in curl: When curl initializes a connection it will call the global NSS_InitContext() method and will specify the certificate db as 'sql:/etc/pki/nssdb', so the system default keystore will be used. Due to the implementation details of NSS's init method only the first call will have its arguments honored.

Understandably, RPM will want to make sure that users get a consistent experience, no matter what plugin/etc calls init().

A side effect of this is that curl's call to NSS_InitContext() is ignored and no keystore will be available to the process.- For an end user this means they effectively can't use EC and several other client certificates when connecting to a YUM repository. 

Additionally, they are left with very little debug-ability, where curl will work:
```
[root@localhost log]# curl  -v  --cacert /usr/share/mycompany/my-certificate-authority.pem  --cert my_cert_in_the_cert_store  https://rpm.mycompany.tld/rhel/x86_64/repodata/repomd.xml
* About to connect() to rpm.mycompany.tld port 443 (#0)
*   Trying XXXXX...
* Connected to rpm.mycompany.tld (XXXX) port 443 (#0)
* Initializing NSS with certpath: sql:/etc/pki/nssdb
*   CAfile: /usr/share/mycompany/my-certificate-authority.pem
  CApath: none
* NSS: using client certificate: my_cert_in_the_cert_store'
```

And yum will not:
```
[root@localhost log]# URLGRABBER_DEBUG="1,/tmp/urlgrab.txt"  yum search mypackage
Loading "fastestmirror" plugin
Config time: 0.006
Yum version: 3.4.3
Setting up Package Sacks
Loading mirror speeds from cached hostfile
 * base: mirrors.ocf.berkeley.edu
 * extras: repos.lax.quadranet.com
 * updates: repos.lax.quadranet.com
* About to connect() to rpm.mycompany.tld port 443 (#12)
*   Trying XXXXX...
* Connected to rpm.mycompany.tld (XXXXX) port 443 (#12)
* Initializing NSS with certpath: sql:/etc/pki/nssdb
* warning: CURLOPT_CAPATH not a directory (/usr/share/mycompany/my-certificate-authority.pem)
*   CAfile: /usr/share/mycompany/my-certificate-authority.pem
  CApath: /usr/share/mycompany/my-certificate-authority.pem
* NSS: client certificate not found: my_cert_in_the_cert_store


........

* SSL connection using TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
> GET /rhel/x86_64/repodata/repomd.xml HTTP/1.1
User-Agent: urlgrabber/3.10 yum/3.4.3
Host: rpm.mycompany.tld
Accept: */*

* The requested URL returned error: 400 Bad Request
* Closing connection 12
https://rpm.mycompany.tld/rhel/x86_64/repodata/repomd.xml: [Errno 14] HTTPS Error 400 - Bad Request
Trying other mirror.

........

failure: repodata/repomd.xml from mycompany-repo: [Errno 256] No more mirrors to try.
https://rpm.mycompany.tld/rhel/x86_64/repodata/repomd.xml: [Errno 14] HTTPS Error 400 - Bad Request
[root@localhost log]# 
```

**Comment by pmatilai** [comment]
NSS_InitContext() is supposed to return a *private* NSS context (see https://wiki.mozilla.org/NSS_Library_Init), and rpm uses it on NSS versions where available. Either rpm was compiled against an NSS version that doesn't have that function (in which case this is a distro issue) or  the context is not as private as the docs seem to indicate, in which case it's an NSS bug. 

Sorry but there's simply nothing rpm can do about either problem. It's just not possible to try co-operate with arbitrary other consumers of NSS with different and potentially conflicting configuration. That is precisely the problem  NSS_InitContext() is supposed to solve over the older NSS_Init()/NSS_NoDB_Init() and if it doesn't, it's an NSS bug.

**Comment by jhlange** [comment]
Fair that this is a NSS nuisance. I'll try opening a bug there. It is unlikely to be fixed, because this is going to be a big behavior change on their end. Sadly, much of the state is global. 

For me, I've hacked this by making a global python sitecustomization.py where I override your default module loading behavior for rpm and yum bindings, and pass a location where I trust the nssdb certs.... It isn't a great hack for us to ship to customers, but it will be good enough for my use case and works fine with yum.
